File: packages/julia/src/Seguranca/Ataques/AtaqueMac.jl
"""
Ataque ao MAC: tenta truncar o campo de integridade, zerá-lo e forçar (força
bruta de 1 byte) valores para ver se algum token adulterado é aceito. Com um
MAC de 32 bytes, nenhuma tentativa deveria passar.
"""
struct AtaqueMac <: Ataque
    mensagem::String
end

AtaqueMac(; mensagem::AbstractString = "mensagem para o ataque de MAC") =
    AtaqueMac(String(mensagem))

nome(::AtaqueMac) = "Força bruta e truncamento do MAC"

function executar(a::AtaqueMac, alvo::AlvoCriptografico)
    pref = prefixo(alvo)
    token = encrypt(alvo, a.mensagem)
    decodificado = base64url_decode(alvo, token[length(pref)+1:end])
    campos = decompor(alvo, decodificado)

    montar(integridade) = pref * base64url_encode(
        alvo,
        recompor(alvo, (integridade=integridade, ciphertext=campos.ciphertext, iv=campos.iv)),
    )

    aceitos = String[]

    # 1) integridade zerada
    try
        decrypt(alvo, montar(zeros(UInt8, length(campos.integridade))))
        push!(aceitos, "integridade zerada")
    catch
        # esperado
    end

    # 2) integridade truncada pela metade
    try
        decrypt(alvo, montar(campos.integridade[1:16]))
        push!(aceitos, "integridade truncada (16 bytes)")
    catch
        # esperado
    end

    # 3) força bruta de 1 byte do MAC (255 variantes; pula o valor original,
    # que reconstruiria o próprio token válido e não é uma forja)
    base = copy(campos.integridade)
    for v in 0:255
        if v == base[1]
            continue
        end
        tentativa = copy(base)
        tentativa[1] = UInt8(v)
        try
            decrypt(alvo, montar(tentativa))
            push!(aceitos, "byte 0 do MAC = $v")
        catch
            # esperado
        end
    end

    vulneravel = !isempty(aceitos)

    return ResultadoAtaque(
        nome(a),
        vulneravel,
        vulneravel ? "critica" : "info",
        vulneravel ?
        "$(length(aceitos)) variante(s) de MAC aceitas: $(join(aceitos[1:min(5, length(aceitos))], "; "))" :
        "Nenhuma das 257 variantes (zerada, truncada, 255 bytes forçados) foi aceita",
        Dict{String,Any}("aceitos" => aceitos),
    )
end

File: packages/python/src/Seguranca/Ataques/AtaqueMac.py
"""Ataque ao MAC: tenta truncar o campo de integridade, zerá-lo e forçar
(força bruta de 1 byte) valores para ver se algum token adulterado é aceito.
Com um MAC de 32 bytes, nenhuma tentativa deveria passar.
"""

from ..ResultadoAtaque import ResultadoAtaque


class AtaqueMac:
    def __init__(self, mensagem: str = "mensagem para o ataque de MAC"):
        self.mensagem = mensagem

    def nome(self) -> str:
        return "Força bruta e truncamento do MAC"

    def executar(self, alvo) -> ResultadoAtaque:
        token = alvo.encrypt(self.mensagem)
        decodificado = alvo.base64url_decode(token[len(alvo.prefixo()) :])
        campos = alvo.decompor(decodificado)

        def montar(integridade: bytes) -> str:
            return alvo.prefixo() + alvo.base64url_encode(
                alvo.recompor(
                    {
                        "integridade": integridade,
                        "ciphertext": campos["ciphertext"],
                        "iv": campos["iv"],
                    }
                )
            )

        aceitos = []

        # 1) integridade zerada
        try:
            alvo.decrypt(montar(bytes(len(campos["integridade"]))))
            aceitos.append("integridade zerada")
        except Exception:
            # esperado
            pass

        # 2) integridade truncada pela metade
        try:
            alvo.decrypt(montar(campos["integridade"][:16]))
            aceitos.append("integridade truncada (16 bytes)")
        except Exception:
            # esperado
            pass

        # 3) força bruta de 1 byte do MAC (255 variantes; pula o valor original,
        # que reconstruiria o próprio token válido e não é uma forja)
        base = campos["integridade"]
        for v in range(256):
            if v == base[0]:
                continue
            tentativa = bytearray(base)
            tentativa[0] = v
            try:
                alvo.decrypt(montar(bytes(tentativa)))
                aceitos.append(f"byte 0 do MAC = {v}")
            except Exception:
                # esperado
                pass

        vulneravel = len(aceitos) > 0

        return ResultadoAtaque(
            self.nome(),
            vulneravel,
            "critica" if vulneravel else "info",
            (
                f"{len(aceitos)} variante(s) de MAC aceitas: {'; '.join(aceitos[:5])}"
                if vulneravel
                else "Nenhuma das 257 variantes (zerada, truncada, 255 bytes forçados) foi aceita"
            ),
            {"aceitos": aceitos},
        )
